synapse/storage/databases/main/events_worker.py

```diff
@@ -485,20 +505,48 @@ async def _get_events_from_cache_or_db(self, event_ids, allow_rejected=False):
 
         Args:
 
-            event_ids (Iterable[str]): The event_ids of the events to fetch
+            event_ids: The event_ids of the events to fetch
 
-            allow_rejected (bool): Whether to include rejected events. If False,
+            allow_rejected: Whether to include rejected events. If False,
                 rejected events are omitted from the response.
 
         Returns:
-            Dict[str, _EventCacheEntry]:
-                map from event id to result
+            map from event id to result
         """
         event_entry_map = self._get_events_from_cache(
-            event_ids, allow_rejected=allow_rejected
+            event_ids,
         )
 
-        missing_events_ids = [e for e in event_ids if e not in event_entry_map]
+        missing_events_ids = {e for e in event_ids if e not in event_entry_map}
+
+        # We now look up if we're already fetching some of the events in the DB,
+        # if so we wait for those lookups to finish instead of pulling the same
+        # events out of the DB multiple times.
+        already_fetching: Dict[str, defer.Deferred] = {}
+
+        # We also add entries to `self._current_event_fetches` for each event
+        # we're going to pull from the DB. We use a single deferred that
+        # resolves to all the events we pulled from the DB (this will result in
+        # this function returning more events than requested, but that can
+        # happen already due to `_get_events_from_db`).
+        fetching_deferred = ObservableDeferred(defer.Deferred())
+
+        for event_id in missing_events_ids:
+            deferred = self._current_event_fetches.get(event_id)
+            if deferred is not None:
+                # We're already pulling the event out of the DB, ad the deferred
+                # to the collection of deferreds to wait on.
+                already_fetching[event_id] = deferred.observe()
+            else:
+                # We're not already pulling the event from the DB, so add our
+                # deferred to the the map of events that are being fetched.
+                self._current_event_fetches[event_id] = fetching_deferred
+                fetching_deferred.observe().addBoth(
+                    lambda _, event_id: self._current_event_fetches.pop(event_id, None),
```

Comment:
why not do this for all the `missing_events_ids` before calling `fetching_deferred.callback` at line 570?